### DISCUSSION TITLE
How to get User After Confirm Verification

### DISCUSSION THREAD

**Discussion by Arinji2** [pocketbase/pocketbase#3549] [Q&A] [answered]
Hiya, so currently im using the SSR version of pocketbase (no authstore, just put the data in cookies and get from there on all requests) now the flow is usually
`/register -> /login (cookie added) -> Verified -> Dash || Not Verified -> /verify -> /login -> /dash`
The reason you need to login again after verify is cause i dont know how to get the user data again after confirm verification.. but this is just a long login flow..  is there any way for me to get the user data after confirm verification which i can yeet into the cookies again?
     
   https://github.com/Arinji2/vibeify

**Comment by ganigeorgiev** [comment]
If you really need the auth data on confirm verification you can override the default response using the [`onRecordAfterConfirmVerificationRequest` hook](https://pocketbase.io/docs/js-event-hooks/#onrecordafterconfirmverificationrequest). Here is an example using the JS `pb_hooks`:

```js
// pb_hooks/main.pb.js

onRecordAfterConfirmVerificationRequest((e) => {
    return $apis.recordAuthResponse($app, e.httpContext, e.record, null);
})
```

_edit: replaced onRecordAfterRequestVerificationRequest with onRecordAfterConfirmVerificationRequest_

**Comment by Arinji2** [comment]
I was thinking more like a way to get the user data directly? Like something like get data of user as an admin?

**Reply by ganigeorgiev** [reply]
In that case you can send a regular `getOne()` request against the user record. The user record id can be extracted from the token payload. For example:

```js
import PocketBase, { getTokenPayload } from "pocketbase";

// authenticate as admin (after this call pb.authStore now has admin token)
await pb.admins.authWithPassword("test@example.com", "1234567890")

// extract the id from the verification token
const payload = getTokenPayload(yourVerificationToken);

// retrieve the user data
const record = await pb.collection(payload.collectionId).getOne(payload.id);
```

**Reply by Arinji2** [reply]
beautiful, awesome mate..thanks :DD

**Reply by Arinji2** [reply]
Hiya so i was trying it out, when i realized i was not getting the token for the user, for my login form i set this as the cookie
``
  const { token, record: model } = await pb
      .collection("users")
      .authWithPassword(parsedForm.data.user, parsedForm.data.password);

    const cookie = JSON.stringify({ token, model });

    cookies().set("pb_auth", cookie, {
      secure: true,
      path: "/",
      sameSite: "strict",
      httpOnly: true,
    });

    
But when i tried out your suggestion, i saw the record is only the user model, and the payload does not have a user token.. how do i like make a new token for the user?

**Reply by ganigeorgiev** [reply]
I'm failing to understand what you are trying to do, but there is no builtin endpoint at the moment that will return a token for any user.

For "impersonate" example you can check https://github.com/pocketbase/pocketbase/discussions/3401.

**Reply by Arinji2** [reply]
Basically how my auth works is that when someone logs in I put their token and auth model in the cookies and on every protected page use those...but sinc you don't get the user when you confirm verification I can't get the latest user data unless I force them to login which is bad ux... The suggestion you provided works but it only gives the auth model... I also need the token. I was thinking that since I have the old token, could I just use that with the new user data? Since verify requires a prior login

**Reply by ganigeorgiev** [reply, marked as answer]
Sorry, I'm even more confused now.

In any case if you want to retrieve the token and user data as admin for any other user, please check the above "impersonate" example in https://github.com/pocketbase/pocketbase/discussions/3401.

But again, I don't understand what you are trying to do and why the initial suggestion in https://github.com/pocketbase/pocketbase/discussions/3549#discussioncomment-7318174 doesn't work for you (_the overwritten auth response will contain both the token and user data_), but I'll not pursue it further.

**Comment by bloor** [comment]
What I ended u doing, was storing into localstorage/cookie (your choice) the users's email+pass from the REGISTRATION FORM. And retrieve them on the next "aerify-email" trigger/page.
**Don't forget to cleanup the storage.**

```JS
	async function submitForm() {
		await $pb.collection(`users`).create({
			username: username.value,
			email: email.value,
			password: password.value,
			passwordConfirm: passwordConfirm.value,
		}).then(() => {
			// account created - and email sent
			localStorage.setItem("pb_login_email", email.value)
			localStorage.setItem("pb_login_password", password.value)
		}).catch((e) => {
                        //BIG ERROR WARNING
		})
	}
```

And then, on verification page/script/endpoint:
```JS
	try {
		await $pb.collection(`users`).confirmVerification(route.query?.token)
		await $pb.collection(`users`).authWithPassword(
			localStorage.getItem("pb_login_email"),
			localStorage.getItem("pb_login_password")
		)
		
		// CLEAR THE STORAGE 
              localStorage.removeItem("pb_login_email"),
              localStorage.removeItem("pb_login_password")

		navigateTo('/');
	} catch (e) {
		showError({
			statusCode: 400,
			statusMessage: e.data?.data?.token ? e.data.data.token.message : e.message,
		})
	}
```
